toolkit/Yara.cpp

```diff
@@ -0,0 +1,290 @@
+                        return -1;
+                    case ERROR_CALLBACK_ERROR:
+                        DARWIN_LOG_ERROR("Toolkit::YaraEngine::ScanData:: could not scan memory -> callback error");
+                        return -1;
+                    case ERROR_TOO_MANY_MATCHES:
+                        DARWIN_LOG_ERROR("Toolkit::YaraEngine::ScanData:: could not scan memory -> too many matches");
+                        return -1;
+                    default:
+                        DARWIN_LOG_ERROR("Toolkit::YaraEngine::ScanData:: could not scan memory -> unknown error");
+                        return -1;
+                }
+            }
+
+            if(_rule_match_list.size() > 0) {
+                certitude = 100;
+                return 1;
+            }
+
+            certitude = 0;
+            return 0;
+        }
+
+        rapidjson::Document YaraEngine::GetResults() {
+            const char *yaraRuleTag;
+            rapidjson::Document rules;
+            rapidjson::Document::AllocatorType& alloc = rules.GetAllocator();
+
+            if(_rule_match_list.size() > 0) {
+                rules.SetArray();
+
+                for(auto rule : _rule_match_list) {
+                    rapidjson::Value ruleJson(rapidjson::kObjectType);
+                    rapidjson::Value tags(rapidjson::kArrayType);
+                    ruleJson.AddMember("rule", rapidjson::StringRef(rule->identifier), alloc);
+
+                    yr_rule_tags_foreach(rule, yaraRuleTag) {
+                        tags.PushBack(rapidjson::StringRef(yaraRuleTag), alloc);
+                    }
+
+                    ruleJson.AddMember("tags", tags, alloc);
+                    rules.PushBack(ruleJson, alloc);
+                }
+            }
+            else {
+                rules.SetNull();
+            }
+
+            return rules;
+        }
+
+        int YaraEngine::ScanOrImportCallback(int message, void *messageData, void *userData) {
+            DARWIN_LOGGER;
+            YR_RULE *rule;
+
+            YaraEngine *engine = (YaraEngine *)userData;
+
+            switch(message) {
+                case CALLBACK_MSG_RULE_MATCHING:
+                    rule = (YR_RULE *)messageData;
+                    engine->AddRuleToMatch(rule);
+                    DARWIN_LOG_INFO("Toolkit::YaraEngine::ScanOrImportCallback:: rule match");
+                    break;
+                case CALLBACK_MSG_RULE_NOT_MATCHING:
+                    break;
+                case CALLBACK_MSG_SCAN_FINISHED:
+                    break;
+                case CALLBACK_MSG_IMPORT_MODULE:
+                    break;
+                case CALLBACK_MSG_MODULE_IMPORTED:
+                    break;
+            }
+
+            return CALLBACK_CONTINUE;
+        }
+
+        void YaraEngine::AddRuleToMatch(YR_RULE *rule) {
+            _rule_match_list.insert(rule);
+        }
+    } // namespace toolkit
+} // namespace darwin
+
+
+namespace darwin {
+
+    namespace toolkit {
```

Comment:
No need to close and reopen the namespaces.

```suggestion
```

Reply:
Done in 3e21146